### ISSUE TITLE
Dependencies wrongly put on class-path rather than module-path

### ISSUE THREAD

**Issue by desruisseaux** [gradle/gradle#25954]
### Expected Behavior

When invoking Java tools such as `java`, `javac` or `javadoc`, the project dependencies can be put either on the class-path or on the module-path using the command-line `--class-path` and `--module-path` options respectively. Maven and Gradle use automatically the module-path if all the following conditions are true:

1. the dependency is modularized (i.e. contains a `module-info.class` file or an `Automatic-Module-Name` attribute in `MANIFEST.MF`), and
2. the project using the dependency is itself modularized.

Condition 1 is okay in most cases (but not always), but 2 is problematic. The fact that a dependency is declared on the class-path rather than the module-path changes the way that `java.util.ServiceLoader` discovers the provided services.

* If the dependency is on the class-path, `ServiceLoader` scans the content of `META-INF/services` directory.
* If the dependency is on the module-path, `ServiceLoader` uses the declarations in `module-info.class`.

Even if condition 2 is false (i.e. a project is not modularized), modularized dependencies still need to be declared on the module-path _for allowing the dependency to discover its own services, or the services of a transitive modularized dependency._ If a modularized dependency is put on the class-path instead, it has consequence not only for the project using that dependency, **but also for the dependency itself, which become unable to use its own `module-info.class`.**

# Demonstration
The attached test case contains two trivial Gradle sub-projects, named `service` and `client`. The first Gradle sub-project declares a dummy services with 4 providers, named _A_, _B_, _C_ and _D_. Providers _A_ and _D_ are declared in `module-info`. Providers _B_ and _C_ are declared in `META-INF/services`. The latter directory should not exist anymore in a JPMS module, but is added in this test for demonstration purposes. A `ShowMyServices` class lists the services discovered by `java.util.ServiceLoader`.

The second Gradle sub-project has only a `main` method invoking `ShowMyServices`. This second module intentionally has no `module-info.java` file. The use case is a big module that we cannot modularize immediately (because modularization brings stronger encapsulation, which may require API changes in the project to modularize), but still want to use modularized dependencies. The expected output is:

```
Start searching for services...
Provider A declared in module-info.
Provider D declared in module-info.
Done.
The dependency has been loaded as named module. Great!
This is what we need for the `module-info` to be used.
```

### Current Behavior

Running the test case with Gradle 8.2.1 show the following output:

```
Start searching for services...
Provider B declared in META-INF.
Provider C declared in META-INF.
Done.
The dependency has been loaded as an unnamed module.
Consequently its `module-info` file has been ignored,
and the `META-INF/services` directory is used instead.
```

Because `service.jar` has been put on the class-path rather than module-path, it has been loaded by the JVM as an unnamed module despite `service.jar` being a JPMS module. Consequently its `module-info.jar` is ignored. The problem is that `service.jar` became an unnamed module not only in the eyes of `client.jar` (which could be understandable since `client.jar` is unmodularized), but more important it became unnamed in the eyes of `services.jar` itself, potentially breaking assumptions made by the library developer. In particular, all services provided by the library become unavailable unless the library developer provided compatibility workaround by copying `module-info` content to `META-INF/services/`.

### Context (optional)

We are trying to modularize the [Apache Spatial Information System (SIS)](https://sis.apache.org/) project. While it is possible to do JPMS to some extent with Gradle and Maven as they stand today, taking full advantage of JPMS requires that we reorganize source code, migrating from _Package Hierarchy_ to _Module Source Hierarchy_. Not many developers are aware of the existence of the latter, maybe because neither Gradle or Maven support it. However thanks to Gradle flexibility, supporting _Module Source Hierarchy_ in Gradle is difficult but not impossible. The process is described here:

* [Java Module Source Hierarchy in a Gradle sub-project](https://geomatys.github.io/draft/Modularization.html)
  * Overview of directory layout change
  * Apache SIS restructuring
  * Gradle configuration for Module Source Hierarchy
    * Gradle issues and workarounds
  * Advantages of Source Module Hierarchy
  * Limitations of Source Module Hierarchy
  * Quasi-blocker Maven bug
  * Conclusion
    * Ideas for Gradle evolution

The quasi-blocker Maven bug is the issue described here, and has also been reported to Maven as [MNG-7855](https://issues.apache.org/jira/browse/MNG-7855). The issue is less a blocker with Gradle because of its flexibility, but the current workarounds are still unsatisfying. The bug is that Gradle modularisation is trying to do black magic for deciding whether to put a dependency on the class-path or module-path, but this black magic is wrong at least in some important scenarios (e.g. gradual modularisation of large projects) and I found no easy easy way to override it with the correct behaviour.

# Proposed solution
We need a way to control whether to put each dependency on the class-path or module-path. This is a critical requirement, just improving the current black magic is not sufficient. However improving that black magic could be a short-term solution, and may still be useful in long term as a default behavior. The improvement could be:

* Drop condition 2 completely (I think it would be the best behavior).
* If condition 2 is kept for compatibility reason, makes it able to recognize _Module Source Hierarchy_ in addition of _Package Hierarchy_.

In longer term, I propose to extend the way that we specify dependencies in Gradle for allowing us to specify explicitly how to handle a dependency in JPMS context. For example:

```xml
dependencyResolutionManagement {
    versionCatalogs {
        create("libs") {
            library("for", "bar", "foo") {
                version {
                    prefer("1.0")
                }
                JPMS {
                  type("module")   // Other allowed values: "unnamed" or "auto" (the default)
                }
            }
        }
    }
}
```

In above example, JPMS is used only for declaring whether to put the dependency in class-path or module-path. But I propose to add such block because in the future, it could be expanded with more elements for controlling other options such as `--add-exports`. But that would be the topic of a separated discussion.

### Steps to Reproduce

Unzip this small test case (8 kb):

[GradleModulepathBug.zip](https://github.com/gradle/gradle/files/12209282/GradleModulepathBug.zip)

Run `gradle test`. The output show that `module-info` has been ignored and `META-INF/services` used instead.

Run `gradle jar`, then reproduce Gradle behavior on the command-line:

```
java --class-path service/build/libs/service.jar:client/build/libs/client.jar test.client.Main
```

Run the following command-line for seeing the expected behavior. Note that the difference compared to above is that the `service.jar` dependency is now put on the module-path instead of the class-path.

```
java --module-path service/build/libs/service.jar --class-path client/build/libs/client.jar --add-modules ALL-MODULE-PATH test.client.Main
```

### Gradle version

8.2.1

**Comment by ov7a** [comment]
Thank you for your interest in Gradle!

This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

***

There is a [plugin](https://github.com/gradlex-org/extra-java-module-info) that might have the functionality you want.

**Comment by desruisseaux** [comment]
Thanks @ov7a for the link to the plugin, I wasn't aware of it. However its documentation said: _"This plugin allows you to add module information to a Java library that does not have any"_, which is not the case here (we do have `module-info` in every modules). The issue here is that:

* Gradle fails to see that a `module-info` exists in some cases (e.g. when using _Module Source Hierarchy_), and consequently wrongly thinks that our project is not modularized.
* Even if the above point was fixed, the issue is not only in our project. The issue can impact any project using our library. Or more generally, any non-modularized application using a modularized library. They are impacted by questionable black magic about what to put on the module-path.

The idea that a non-modularized project must put all its dependencies on the class-path seems to be a widely-spread misunderstanding of how JPMS work. There is possibly different opinions about what is best (and it may depend on the library), but the current problem is that Gradle imposes its black magic with no way I can see to specify what we need.

**Comment by jjohannes** [comment]
@desruisseaux if I understand your example correctly, the reason why "Gradle fails to see that a module-info exists" is that if the **source set (!)** you are compiling/running is itself not a module, then Gradle treats everything as classpath. You can make your tests a module (blackbox testing) by adding a `client/src/test/java/module-info.java`. Then it should work as you expect.

If you can't do this for some reason and want to do so called "whitebox testing", where the tests run on the module path without separate module-info, we also have a plugin for that: https://github.com/gradlex-org/java-module-testing

From my perspective, if there is anything that could potentially be improved in Gradle core, it is for the whitebox testing case. The plugin needs to do some things that are a bit involved to make it work (but it works).

I also recently did a video on the topic. Maybe it helps a bit to clarify things: https://youtu.be/6rFEDcP8Noc

**Comment by desruisseaux** [comment]
Hello @jjohannes, thanks for your reply.

> (…snip…) the reason why "Gradle fails to see that a module-info exists" is that if the source set (!) you are compiling/running is itself not a module, then Gradle treats everything as classpath.

On the "fails to see that a module-info exists" point, I'm thinking about a project using _Module Source Hierarchy_, which is the subject of #25974. Gradle is currently not designed for seeing the `module-info` files in that case, with no workaround I could find.

On the "Gradle treats everything as classpath" point, this fact was already stated in the preamble of this issue description as point 2 below the first paragraph. As this issue tries to explain, this Gradle (and Maven) behavior is problematic. Allowing developers to change this behavior is what this issue is asking.

The decision to treat everything as classpath when condition 2 is true seems to be based on a wrong understanding of how JPMS work. The same misunderstanding [exists on Maven side](https://lists.apache.org/thread/w1cxh2r5l41zqcbftybk1wcjg4x8vrld) as well. The decision seems to be based on the belief that if an application has no `module-info`, then all its dependencies *must* be on the classpath (otherwise Gradle or Maven would have provided some configuration options). This is not true, and this is what the attached example demonstrates. The intend of that example is not only to demonstrate that Gradle doesn't work as expected. The main intend is to demonstrate **how it should work**, with `java` invoked on the command line with the right parameters. What this issue is asking is a way to get Gradle to invoke *all* Java tools (`java`, `javac`, `javadoc`, _etc._) with the right _class-path versus module-path_ parameters. There is possibly different opinions about what is right, but the example shows that a non-modularized client application can use module path for its modularized dependencies, that it works and is legitimate. The difficulty in doing so is caused by Gradle/Maven questionable black magic, not by JPMS constraint.

> You can make your tests a module (blackbox testing) by adding a client/src/test/java/module-info.java. Then it should work as you expect.

The omission of that `module-info.java` file was intentional, because it is the whole point of this example. The use cases are:

* A big or complex client application that cannot be modularized right now.
* A client application that we do not control, e.g. all consumers of our library deployed on Maven Central.

Gradle or Maven should not block a non-modularized application to use modularized dependencies. Putting all dependencies on the classpath, with no way for developers to control that, can *block* usage of modularized dependencies. The dependencies behave differently depending on whether they were specified on the class-path or on the module-path. Not only `java.util.ServiceLoader` do not behave in the same way (which is what the example demonstrate), but numerous OpenJDK methods behave differently, especially the methods related to reflection.

> If you can't do this for some reason and want to do so called "whitebox testing", where the tests run on the module path without separate module-info, we also have a plugin for that

As said above, I do add `module-info` but Gradle can't see them because of _Source Module Hierarchy_. I didn't reproduced that problem in the attached example because it is not the topic of this issue, so I omitted the `module-info` instead.

Testing is not the issue. Gradle is flexible enough for allowing us to add the `--patch-module` and `--add-exports` options needed (it could be made easier, but this is another topic). The issue is about allowing non-modularized client applications to use our library. We do not control user's code, we can't force them to modularize their own projects if they are not ready to do so. If they don't modularize, the fact that Gradle and Maven refuse to put the library on the module-path means that we have to workaround with duplication (`module-info` services copied into `META-INF/services/` directory), must refrain from using features such as service singletons (because they are not supported on `META-INF/services/`), and are at risk of behaviorial differences even if we try to be careful, because OpenJDK methods that behave differently can manifest themselves in unexpected places.

**Comment by jvandort** [comment]
So from what I understand, it seems like solving both of these below issues would fix your problem. Is that correct?
* Include a way to disable (or by default, disable) condition 2 (`the project using the dependency is itself modularized.`)
* Fix https://github.com/gradle/gradle/issues/25974, at least partially, by supporting Module Hierarchy when detecting if a dependency is a module or not (when searching for `module-info.class` file or an `Automatic-Module-Name`)
    * This would require [JavaModuleDetector](https://github.com/gradle/gradle/blob/1045229ba88d7165101cba22d99f5ad1465bcd81/subprojects/core/src/main/java/org/gradle/internal/jvm/JavaModuleDetector.java#L38) to be updated

**Comment by desruisseaux** [comment]
Hello @jvandort. Yes exactly, thanks! The first point (disable condition 2) is the most important, as it impacts all non-modularized projects (not only the projects wanting to use Module Source Hierarchy). Disabling it by default would be the right thing to do in my opinion, but has two backward compatibility drawbacks:

* Some modularized dependencies may behave differently. They should behave more like intended by the dependency developer, but users may be surprised by behavioural changes.
* Applications may not work anymore if the `--add-modules ALL-MODULE-PATH` option (or something equivalent) is not added to the `java` command.

Disabling condition 2 is probably the easiest short-term solution, and would fix my problem. For a longer term solution, it would be nice to have a control on a case-by-case basis by extending the syntax of `dependencyResolutionManagement` block or something equivalent. There is many JPMS-related options that could go there:

* Whether to put the dependency on class-path or module-path (this issue).
* Whether to declare the dependency in the `--add-modules` option (more fine-grained than `ALL-MODULE-PATH`).
* Whether to generate automatically `--add-exports` to that dependency (useful for JUnit).
* etc.

**Comment by jvandort** [comment]
I would like to focus our efforts on dealing with condition 2 for now.

I think there is value in exposing an API that gives users more control over how individual dependencies are handled in terms of JPMS. However the exact manner in which this is implemented is certainly up for debate. 

* Should these rules be applied across the entire build (as `dependencyResolutionManagement` would do), or should they be configurable on a project-by-project basis (likely). I can imagine one application requiring an `--add-exports` for a given dependency while another does not. 
* Should these rules be tied to the _dependencies_ themselves? Or perhaps the configurations they are declared on? Or, perhaps set at a component-level by declaring them on the `java` extension. 
* What does the API look like? What needs to be configurable by users? 

The answers to these questions, and many more, will need to be answered before we move forward with any user-facing API changes. The standard process is to write a formal design specification, which is then reviewed by the Gradle Build Tool Team, and is then implemented. 

This requires a significant amount of work, and while I am not necessarily _denying_ we should make an effort here, I doubt that we (the JVM team at Gradle) will be able to focus significant efforts here, given that our current focus is on improving Testing and Publishing. 

This does seem like a significant enough issue that I think we should make progress on it, but I think focusing our effort on solutions that provide a workable solution sooner rather than later is more likely to produce results. 

Given that, in terms of moving forward, I think we should revise this issue, or open a new issue, which focuses solely on the `condition 2` problem, and allow the larger API discussion to happen independently. We usually allow the community to help drive interest in these issues by upvoting them. 

The largest issue I see here is backwards compatibility, which we take quite seriously. I see two options as a solution:
1. Disable condition 2 by default
    a. How does this affect backward compatibility? What behavior changes do you foresee (one obvious one being the ServiceLoader behavior). 
    b. What would we need to do limit behavior changes? Can we automatically add flags to limit the impact of this change? Which ones? Where? You mention some options above, though I am curious if you could our potential options more concretely, as you seem quite familiar working with JPMS in general. 
    c. I can see a number of projects getting burned by this change, for example if they are depending on internals of a modularized dependency. Unless we can resolve this issue, this kind of change in behavior would be unacceptable for a minor-version change. Behavior changes like this would need to go through a deprecation stage where we first warn users of the change in behavior, then actually make the change in a major version. Furthermore, we would likely need a way to _detect_ if this sort of behavior change would break existing users, so that we can warn them _only_ if they have an offending build. 
2. Add an API for users to disable condition 2
    a. This, IMO, is the fallback option. I would prefer a solution that is as seamless as possible -- similar to how modulepath detection (mostly) works now, where in most use cases, the user doesn't need to know whether a given dependency is placed on the modulepath or classpath.

**Comment by desruisseaux** [comment]
Yes, I realize that the long term solution would be a lot of work. One of my goal was to get the discussion on track. For short term, an API for disabling condition 2 without changing the current default (for compatibility reason) is fine. A possibility would be to change the [ModularitySpec](https://docs.gradle.org/8.2/javadoc/org/gradle/api/jvm/ModularitySpec.html) interface as below:

* Deprecate the `getInferModulePath()` method.
* Replace by a `getModulePathStrategy()` method, which would return an enumeration value.

Enumeration values could be (alternative name proposals are welcome):

* `INFER_IF_MODULARIZED_PROJECT`: the current behavior.
* `MODULARIZED_DEPENDENCIES`: condition 1 only, without condition 2.
* `ALL`: unconditionally put all dependencies on the module path, even if a dependency is not modularized (in which case the JVM derives a module name from the JAR file name).
* `NONE`: unconditionally put all dependencies on the class path, even if modularized. Sometime useful for testing purposes.

> What behavior changes do you foresee (one obvious one being the ServiceLoader behavior).

Behavioural changes include everything related to reflection, with `ServiceLoader` being only one of them. JPMS applies stronger encapsulation not only to the "ordinary" Java code, but also to resources and all usages of reflection methods. For example if a dependency calls `Foo.class.getResourceAsStream("Bar.properties")`, that method always returns a non-null value (assuming that the `Bar.properties` file exists) if the code invoking `Class.getResourceAsStream(…)` is on the class-path. But the exact same method call may return `null` if the code is on the module-path and `Foo` is a class defined in a non-exported package of another module. This restriction is enforced even if `Foo.class` was a parameter specified by the latter module, because `Class.getResourceAsStream(…)` is sensitive to the module of the *caller* in addition to the module of the class.

More generally, there is an internal OpenJDK annotation named `@CallerSensitive` which is applied on OpenJDK methods that behave differently depending on who invoke them. Last time I checked (a few days ago), 167 OpenJDK methods are annotated with `@CallerSensitive`. Not all of them are module-sensitive (`@CallerSensitive` is also used for logging stuff or thread stuff), but many are. All those OpenJDK methods can potentially behave differently depending on whether the JAR file containing the class invoking those methods was declared on the class path or on the module path.

Since the class-path versus module-path decisions can have deep implications on the behaviour of an application, I think they should be under developer's control. I agree that the long term solution (new API in `dependencyResolutionManagement` or something along that line) would require more discussions. In short term, replacing the boolean by an enumeration in `ModularitySpec` should be sufficient for unblocking common cases.

**Comment by Vampire** [comment]
Another option could be to fix the logic, so that JPMS modules are always placed on the module-path and non-modules are always placed on the classpath, but hide the new logic behind a feature preview flag, and then in the next major release make the feature preview permanent and list it as potentially breaking change. Probably logging a deprecation warning if the feature flag is not enabled and the placement would be different with the flag enabled.

**Comment by jjohannes** [comment]
Thanks for all the clarification @desruisseaux and sorry for misunderstanding the issue in my earlier comment (disregard what I wrote there).

My feeling is that this should be kept configurable (at least for the time being). There might be constellations where users still want to use the classpath only, although some libraries have a `module-info.class`.

But I agree that the better default would eventually be to put things with `module-info.class` on the Module Path. So it could be:

- **Now**: Introduce a new flag on `modularity` (which the Java extension and all the tasks have). E.g. `modularity.classpathOnly`. Initial default is `true`.
- **Next major**: Switch the default to `false`

Or going a step further and instead having one setting with different options as suggested here: https://github.com/gradle/gradle/issues/25954#issuecomment-1670442564 by @desruisseaux. Just be careful there what `MODULARIZED_DEPENDENCIES` means.

Or make it the default (as @Vampire suggest) and add a flexible opt-out/opt-in for both module and class path, as talked about here: https://github.com/gradle/gradle/issues/12630#issuecomment-1125105629

**Comment by desruisseaux** [comment]
Hello Johannes

No problem, thanks for your input! Yes I agree that it should be configurable. A `classpathOnly` option is indeed necessary at least for compatibility reasons (some users may depend on current Gradle/Maven behaviour). But a `modulepathOnly` option is also needed when the library is not JPMS (so it should be on the classpath), but the application decided to rely on automatic module name and put a `requires` directive for that JAR in its `module-info`. This is not recommended, but not illegitimate neither and up to applications to decide. This is why I think that an enumeration may be preferable to boolean flags.

The boolean or enumeration proposed here would be only a medium-term solution. A longer term solution would be to let developers control the class-path versus module-path decision on a case-by-case basis, together with other JPMS related options such as `--add-exports` (useful for testing), etc. It could be part of https://github.com/gradle/gradle/issues/25974#issuecomment-1714210250.
